### PR TITLE
[onert] ostream overloads for index types

### PR DESCRIPTION
**Pull request by wateret**
Introduce ostream overloads for index types. Each types of indexes
has its own prefix.

Plus, this commit includes usage fixes.

ONE-DCO-1.0-Signed-off-by: Hanjoung Lee <hanjoung.lee@samsung.com>

## 5+ Reviews to merge

### CODE REVIEW

**Comment by wateret** [comment]
This was discussed (alone 😅 ) before - https://github.sec.samsung.net/STAR/nnfw/issues/9212 . I'd like to introduce this since I have been feeling that sometimes debug log is hard to read. However switching to this way may be even more confusing to others.

So **I WOULD LIKE TO GET 5+ REVIEWS** before merging. I don't want to merge this if there are many objections 😃

**Comment by wateret** [comment]
Before
```
[LoweredGraph] Dump after permutation insertion
[LoweredGraph] Graph Output : 0
[OpSequences] 0]   OpSequence IN(2) -> { 0(Permute:2:0) } -> OUT(0)
[StaticShapeInferer] SubGraph #0
[StaticShapeInferer] Operand #2, Static, shape : {1 2 2 1}
[StaticShapeInferer] Operand #0, Static, shape : {1 2 2 1}
[LIR] Executor generation of Subgraph 0
[LIR] * Permute(Copy)
[LIR]   - Inputs : Input(2)
[LIR]   - Output : Output(0)
[Linear] Final OpSequence
[Linear] * OP_SEQ {controlflow}   OpSequence IN(2) -> { 0(Permute:2:0) } -> OUT(0)
```

After

```
[LoweredGraph] Dump after permutation insertion
[LoweredGraph] Graph Output : #0
[OpSequences] &0]   OpSequence IN(#2) -> { *0(Permute:#2:#0) } -> OUT(#0)
[StaticShapeInferer] SubGraph $0
[StaticShapeInferer] Operand #2, Static, shape : {1 2 2 1}
[StaticShapeInferer] Operand #0, Static, shape : {1 2 2 1}
[LIR] Executor generation of Subgraph 0
[LIR] * Permute(Copy)
[LIR]   - Inputs : Input(#2)
[LIR]   - Output : Output(#0)
[Linear] Final OpSequence
[Linear] * OP_SEQ {controlflow}   OpSequence IN(#2) -> { *0(Permute:#2:#0) } -> OUT(#0)
```

**Review by hseok-oh — APPROVED** [review]
LGTM

**Comment by glistening** [comment]
@wateret 

entity | symbol
--- | ---
OperationIndex | *
OperandIndex | #
IOIndex | @
OpSequenceIndex | &
SubgraphIndex | $

It is hard to remember because there are 5 symbols and I cannot make association between the meanings and symbols. Could you please give me an idea how to remember this association if you have? If not, maybe I need to print out this when I need to interpret debug log.

I would introduce symbol for `OperandIndex` only, and use abbreviation of 2~3 letters for others.
All of 5 entities are index and I abbreviates without `Index`.

entity | symbol
--- | ---
OperationIndex | `op`
OperandIndex | `$` <br><sub>I use `$` rather than `op` because `#` pop up issue or pr in github. </sub>
IOIndex | `io`
OpSequenceIndex | `osq`
SubgraphIndex | `sgp`

Here is your example with this style:

```
[LoweredGraph] Dump after permutation insertion
[LoweredGraph] Graph Output : $0
[OpSequences] ops0]   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
[StaticShapeInferer] sgp0
[StaticShapeInferer] $2, Static, shape : {1 2 2 1}
[StaticShapeInferer] $0, Static, shape : {1 2 2 1}
[LIR] Executor generation of sgp0
[LIR] * Permute(Copy)
[LIR]   - Inputs : Input($2)
[LIR]   - Output : Output($0)
[Linear] Final OpSequence
[Linear] * OP_SEQ {controlflow}   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
```

**Comment by glistening** [comment]
Beyond this PR, I prefer to fixed width prefix for each phases as I did in my tools: It would look better if it is shown like:

```
[LoweredGraph      ] Dump after permutation insertion
[LoweredGraph      ] Graph Output : $0
[OpSequences       ] ops0]   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
[StaticShapeInferer] sgp0
[StaticShapeInferer] $2, Static, shape : {1 2 2 1}
[StaticShapeInferer] $0, Static, shape : {1 2 2 1}
[LIR               ] Executor generation of sgp0
[LIR               ] * Permute(Copy)
[LIR               ]   - Inputs : Input($2)
[LIR               ]   - Output : Output($0)
[Linear            ] Final OpSequence
[Linear            ] * OP_SEQ {controlflow}   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
```

I would limit the prefix size like 12 letters. `StaticShapeInferer` is too lengthy.

```
[LoweredGraph] Dump after permutation insertion
[LoweredGraph] Graph Output : $0
[OpSequences ] ops0]   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
[StShapeInfer] sgp0
[StShapeInfer] $2, Static, shape : {1 2 2 1}
[StShapeInfer] $0, Static, shape : {1 2 2 1}
[LIR         ] Executor generation of sgp0
[LIR         ] * Permute(Copy)
[LIR         ]   - Inputs : Input($2)
[LIR         ]   - Output : Output($0)
[Linear      ] Final OpSequence
[Linear      ] * OP_SEQ {controlflow}   OpSequence IN($2) -> { *0(Permute:$2:$0) } -> OUT($0)
```

**Comment by wateret** [comment]
> It is hard to remember because there are 5 symbols and I cannot make association between the meanings and symbols. Could you please give me an idea how to remember this association if you have a mnemonic? If not, maybe at least I need a print-out of this table. :)

Please take a look at comments in `runtime/onert/core/include/ir/Index.h`. Some are just chosen randomly, though. 😓

**Comment by wateret** [comment]
> entity | symbol
> --- | ---
> OperationIndex | `op`
> OperandIndex | `$` <br><sub>I use `$` instead of `#` (your suggestion) because `#` causes pop up or reference to issue or pr in github. </sub>
> IOIndex | `io`
> OpSequenceIndex | `osq`
> SubgraphIndex | `sgp`

Thank you for your idea. But if we go this way, I would prefer2-3 alphabets for Operand Index too, for consistency.

**Comment by wateret** [comment]
**A single char symbols** vs. **2-3 alphabet letters**

 Which is better?

**Comment by hyunsik-yoon** [comment]
This depends heavily on personal taste... :-)

For me, those frequently displayed (such as operations and tensors) could have one-char symbol to shorten the length of log. (also people will remember the meaning of symbols better since they see the symbol frequently) 
For other non-frequently displayed things, people would not remember them. So full name or easy-to-read abbreviation would be better. 

Anyway, we need to have `how-to-read-log.md`. Few people will know the meaning of `"#3"` or `"$3"` at first glance. Also, we can use the symbols or abbreviations throughout out documentation, code, and comments for integrity.
Likewise, currently we use `"subg"` in our source code for subgraph. Then it might be better to use `subg 3` in log too.

**Comment by glistening** [comment]
@wateret 

I prefer i) 2 ~ 3 alphabet letters except for operand index, ii) a single char symbols, and iii) 2~3 alphabet letters in order.

Because you removed item i), my choice is  ii) a single char symbols.

@hyunsik-yoon I agree. It was the reason I suggested i). 
Plus, `subg` looks better than `sbg`. 4 letters makes it more readable. :)

**Comment by glistening** [comment]
@wateret I managed to create a mnemonic.

```
`$` has S in its symbol, thus it is **S**ubgraphIndex.
`*` is multiplication operator in C, thus it is OperatorIndex.
`&` is `and` in C, thus it can represent plurals, thus it is OpSequence.
```

1. Could you please replace `@` with `%` for IOIndex? `%` has I and o character in symbol.
2. `#` is OperandIndex. Just memorize!

**Comment by periannath** [comment]
How about using `single char symbols` only for `OperationIndex` and `OperandIndex`? And using full name for other indexes.

`OperationIndex` and `OperandIndex` are heavily used in our logs but others are rarely used. Others are usually used with full name. 

#### `OpSequenceIndex` example
- `[Lower] OpSequence#0 is created for NODE#65(Softmax)`
- `[OpSequences] 20]   OpSequence ...`

#### `SubGraphIndex` example
- `[StaticShapeInferer] SubGraph #0`
- `[LIR] Executor generation of Subgraph 0`

I can't find log with `IOIndex`. Rembmbering two symbols (`OperationIndex` and `OperandIndex`) would be better than remembering five symbols.

**Comment by glistening** [comment]
> @wateret
> I prefer i) 2 ~ 3 alphabet letters except for operand index, ii) a single char symbol, and iii) 2~3 alphabet letters in order.
> Because you removed item i), my choice is ii) a single char symbol.
> 

To make it clear, please don't count my comment above as +1 for a single char symbol.

It was an answer for your question.

> A single char symbols vs. 2-3 alphabet letters
>
> Which is better?

I prefer not to change than those two options.

**Comment by wateret** [comment]
@glistening @periannath @hyunsik-yoon Thank you all for your ideas.

It looks like nobody likes a single char symbol for all the index types.

- OperationIndex : An abbreviation or a symbol
- OperandIndex : An abbreviation or a symbol
- IOIndex : An abbreviation
- OpSequenceIndex : An abbreviation
- SubgraphIndex : An abbreviation

I concluded with the above. So please suggest ideas for each.

**Comment by wateret** [comment]
> entity | symbol
> --- | ---
> OperationIndex | `op`
> OperandIndex | `$` <br><sub>I use `$` instead of `#` (your suggestion) because `#` causes pop up or reference to issue or pr in github. </sub>
> IOIndex | `io`
> OpSequenceIndex | `osq`
> SubgraphIndex | `sgp`

I looked this again and changed my mind just using a symbol for OperandIndex. I agree with this. Because I found out that operand indices occur many times in a line, and others aren't. So shorter is better for operand index.

**Comment by glistening** [comment]
@wateret Thank you for introducing only one or two symbols. It is what I want. 

However, as for using abbreviation for other indexes, I prefer descriptive names that can be understood without other documents as @hyunsik-yoon suggested. We may choose to 

i) use more letters (like 4 letters `opsq`, `subg`, `ornd`...), especially I prefer `subg` than `sbg` 

or 

ii) not to abbreviate all.

**Comment by wateret** [comment]
> ii) not to abbreviate all.

I will give it a shot. We may use it first, and see if the full names are inconvenient.

File: runtime/onert/core/include/ir/Index.h
Comment (wateret) [inline runtime/onert/core/include/ir/Index.h:52]: Updated prefixes

**Comment by wateret** [comment]
Updated result snippet

```
[LoweredGraph] Dump after permutation insertion
[LoweredGraph] Graph Output : $0
[OpSequences] OPSEQUENCE0]  IN($2) -> { OPERATION0(Permute:$2:$0) } -> OUT($0)
[EdgeConsistencyChecker] Total Number of errors : 0
[StaticShapeInferer] SUBGRAPH0
[StaticShapeInferer] $2, Static, shape : {1 2 2 1}
[StaticShapeInferer] $0, Static, shape : {1 2 2 1}
[LIR] Executor generation of Subgraph 0
[LIR] * Permute(Copy)
[LIR]   - Inputs : Input($2)
[LIR]   - Output : Output($0)
[Linear] Final OpSequence
[Linear] * OP_SEQ {controlflow}  IN($2) -> { OPERATION0(Permute:$2:$0) } -> OUT($0)
```

**Review by hseok-oh — APPROVED** [review]

**Review by periannath — APPROVED** [review]
LGTM

**Review by glistening — APPROVED** [review]
LGTM

**Review by dayo09 — APPROVED** [review]
LGTM

**Comment by wateret** [comment]
Got 4 out of 5 reviews, but I think it is enough 😄

**Pull request merged**
